brvs-rules/src/main/java/iroha/validation/rules/impl/billing/BillingRule.java

```diff
@@ -0,0 +1,251 @@
+      throw new IllegalArgumentException("RMQ port must be valid");
+    }
+    if (Strings.isNullOrEmpty(rmqExchange)) {
+      throw new IllegalArgumentException("RMQ exchange must not be neither null nor empty");
+    }
+    if (Strings.isNullOrEmpty(rmqRoutingKey)) {
+      throw new IllegalArgumentException("RMQ routing key must not be neither null nor empty");
+    }
+
+    this.getBillingURL = getBillingURL;
+    this.rmqHost = rmqHost;
+    this.rmqPort = rmqPort;
+    this.rmqExchange = rmqExchange;
+    this.rmqRoutingKey = rmqRoutingKey;
+    runCacheUpdater();
+  }
+
+  private void runCacheUpdater() throws IOException {
+    if (isRunning) {
+      return;
+    }
+    isRunning = true;
+    readBillingOnStartup();
+    getMqUpdatesObservable().subscribeOn(Schedulers.from(Executors.newSingleThreadExecutor()))
+        .subscribe(update -> {
+              final BillingInfo currentBillingInfo = cache
+                  .stream()
+                  .filter(entry -> entry.equals(update))
+                  .findAny()
+                  .orElse(null);
+              if (currentBillingInfo == null
+                  || currentBillingInfo.getUpdated().isBefore(update.getUpdated())) {
+                // Will be replaced, see BillingInfo#hashCode
+                cache.add(update);
+              }
+            }
+        );
+  }
+
+  private void readBillingOnStartup() throws IOException {
+    final JsonObject root = jsonParser.parse(executeGetRequest()).getAsJsonObject();
+    for (BillingTypeEnum billingType : BillingTypeEnum.values()) {
+      final String label = billingType.label;
+      cache.addAll(
+          BillingInfo.parseBillingHttpDto(
+              label,
+              gson.fromJson(
+                  root.getAsJsonObject(label),
+                  new TypeToken<HashMap<String, HashMap<String, JsonObject>>>() {
+                  }.getType()
+              )
+          )
+      );
+    }
+  }
+
+  private String executeGetRequest() throws IOException {
+    URL url = new URL(getBillingURL);
+    HttpURLConnection urlConnection = (HttpURLConnection) url.openConnection();
+    final int responseCode = urlConnection.getResponseCode();
+    if (responseCode != 200) {
+      throw new BillingRuleException(
+          "Couldn't request primary billing information. Response code is " + responseCode
+      );
+    }
+
+    BufferedReader in = new BufferedReader(new InputStreamReader(urlConnection.getInputStream()));
+    String inputLine;
+    StringBuilder response = new StringBuilder();
+    while ((inputLine = in.readLine()) != null) {
+      response.append(inputLine);
+    }
+    in.close();
+    return response.toString();
+  }
+
+  private Observable<BillingInfo> getMqUpdatesObservable() {
+    ConnectionFactory factory = new ConnectionFactory();
+    factory.setHost(rmqHost);
+    factory.setPort(rmqPort);
+    final Connection connection;
+    try {
+      connection = factory.newConnection();
```

Comment:
Are you handle somewhere case of lost existent connection?

Reply:
Do you mean to check something like `connection.isOpen()`? Moreover, I don't see any related functionality in Notary repository 🤷‍♀️